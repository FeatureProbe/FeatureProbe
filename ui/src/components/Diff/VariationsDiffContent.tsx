import { FormattedMessage } from 'react-intl';
import { Table } from 'semantic-ui-react';
import { diffType, positionType } from './constants';
import { DiffResult } from './diff';
import { DiffFieldValue } from './DiffFieldValue';
import { renderField, renderFieldsItems } from './renderDiff';
import fieldStyles from './fields.module.scss';
import styles from './VariationsDiffContent.module.scss';

type DiffFieldValue = {
  type: diffType;
  value: string;
};

interface VariationsField {
  name?: string;
  value?: string;
  description?: string;
}

const VariationsModifyDiffItem: React.FC<{ content: DiffResult; type: positionType }> = (props) => {
  const { content, type } = props;

  return (
    <Table.Row className={styles['diff-item']}>
      {renderField(content, type, (map) => {
        const name = map.get('name');
        const value = map.get('value');
        const description = map.get('description');
        return (
          <>
            <Table.Cell className={styles['table-item-name']}>
              {name ? <DiffFieldValue type={name.type} value={name.value as string} /> : 'null'}
            </Table.Cell>
            <Table.Cell>
              {value ? <DiffFieldValue type={value.type} value={value.value as string} /> : 'null'}
            </Table.Cell>
            <Table.Cell>
              {description ? <DiffFieldValue type={description.type} value={description.value as string} /> : 'null'}
            </Table.Cell>
          </>
        );
      })}
    </Table.Row>
  );
};

const VariationsCountDiffItem: React.FC<{
  content: VariationsField;
  diffType: diffType;
  type: positionType;
}> = (props) => {
  const { content, diffType, type } = props;
  const obj = content;
  const empty = (diffType === 'add' && type === 'before') || (diffType === 'remove' && type === 'after');

  return (
    <Table.Row className={`${styles['diff-item']} ${!empty ? fieldStyles[`diff-item-${diffType}`] : ''}`}>
      <Table.Cell className={styles['table-item-name']}>
        {obj.name && !empty && <DiffFieldValue value={obj.name} />}
      </Table.Cell>
      <Table.Cell>{obj.value && !empty && <DiffFieldValue value={obj.value} />}</Table.Cell>
      <Table.Cell>{obj.description && !empty && <DiffFieldValue value={obj.description} />}</Table.Cell>
    </Table.Row>
  );
};

interface VariationsDiffContentProps {
  content: DiffResult;
}

const VariationsDiffContent: React.FC<VariationsDiffContentProps> = (props) => {
  const { content } = props;

  return (
    <div className={fieldStyles['diff-content']}>
      <div className={fieldStyles['before']}>
        <div className={fieldStyles['diff-fields']}>
          <Table basic="very" unstackable size="small">
            <Table.Header className={fieldStyles['table-header']}>
              <Table.Row>
                <Table.HeaderCell className={styles['table-header-name']}>
                  <FormattedMessage id="common.name.lowercase.text" />
                </Table.HeaderCell>
                <Table.HeaderCell>
                  <FormattedMessage id="common.value.text" />
                </Table.HeaderCell>
                <Table.HeaderCell>
                  <FormattedMessage id="common.description.lowercase.text" />
                </Table.HeaderCell>
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {content &&
                renderFieldsItems<VariationsField>(content, 'before', (value, diffType, _, index) => {
                  if (diffType === 'modify') {
                    return <VariationsModifyDiffItem key={index} type="before" content={value as DiffResult} />;
                  } else {
                    return (
                      <VariationsCountDiffItem key={index} diffType={diffType} type="before" content={value as VariationsField} />
                    );
                  }
                })}
            </Table.Body>
          </Table>
        </div>
      </div>
      <div className={fieldStyles['after']}>
        <div className={fieldStyles['diff-fields']}>
          <Table basic="very" unstackable size="small">
            <Table.Header className={fieldStyles['table-header']}>
              <Table.Row>
                <Table.HeaderCell className={styles['table-header-name']}>
                  <FormattedMessage id="common.name.lowercase.text" />
                </Table.HeaderCell>
                <Table.HeaderCell>
                  <FormattedMessage id="common.value.text" />
                </Table.HeaderCell>
                <Table.HeaderCell>
                  <FormattedMessage id="common.description.lowercase.text" />
                </Table.HeaderCell>
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {content &&
                renderFieldsItems<VariationsField>(content, 'after', (value, diffType, _, index) => {
                  if (diffType === 'modify') {
                    return <VariationsModifyDiffItem key={index} type="after" content={value as DiffResult} />;
                  } else {
                    return (
                      <VariationsCountDiffItem key={index} diffType={diffType} type="after" content={value as VariationsField} />
                    );
                  }
                })}
            </Table.Body>
          </Table>
        </div>
      </div>
    </div>
  );
};

export default VariationsDiffContent;
