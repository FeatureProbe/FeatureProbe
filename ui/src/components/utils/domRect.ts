export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}