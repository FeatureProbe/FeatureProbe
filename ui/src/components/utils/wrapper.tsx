import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
