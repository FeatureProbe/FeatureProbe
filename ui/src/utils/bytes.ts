export function getBytes(str: string) {
  let bytes = str.length,
    i = 0;
  for (; i < bytes; i++) if (str.charCodeAt(i) > 255) bytes++;

  return bytes;
}

const MB = 1048576;
export const MAX_SIZE = MB;