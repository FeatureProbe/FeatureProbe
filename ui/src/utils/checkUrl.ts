function escapeStringRegexp(string: string): string {
  if (typeof string !== 'string') {
    throw new TypeError('Expected a string');
  }

  return string
    .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
    .replace(/-/g, '\\x2d');
}

export function matchUrl(matcher: string, originUrl: string, targetUrl: string): boolean {
  let href = '';
  let hash = '';
  let search = '';

  try {
    const fullUrl = new URL(targetUrl, '');
    href = fullUrl.href;
    hash = fullUrl.hash;
    search = fullUrl.search;
  } catch (error) {
    href = targetUrl;
  }

  let regex;
  let testUrl;

  switch (matcher) {
    case 'EXACT':
      testUrl = href;
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SIMPLE':
      testUrl = href.replace(hash, '').replace(search, '');
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SUBSTRING':
      testUrl = href.replace(search, '');
      regex = new RegExp('.*' + escapeStringRegexp(originUrl) + '.*$');
      break;
    case 'REGULAR':
      testUrl = href.replace(search, '');
      regex = new RegExp(originUrl);
      break;
    default:
      return false;
  }

  return regex.test(testUrl);
}