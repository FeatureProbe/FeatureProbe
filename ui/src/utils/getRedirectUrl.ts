import { getProjectList } from 'services/project';
import { IProject, IEnvironment } from 'interfaces/project';
import { IDictionary } from 'interfaces/targeting';
import { getFromDictionary } from 'services/dictionary';
import { LAST_SEEN } from 'constants/dictionary_keys';

const getInfo = async () => {
  const res =  await getFromDictionary<IDictionary>(LAST_SEEN);
  const { success, data } = res;
  if (success && data) {
    return JSON.parse(data.value);
  } else {
    return {};
  }
};

export const getRedirectUrl = async (defaultRedirectUrl: string) => {
  let redirectUrl = defaultRedirectUrl;
  const info = await getInfo();
  const { projectKey, environmentKey } = info;

  if (!!projectKey && !!environmentKey) {
    redirectUrl = `/${projectKey}/${environmentKey}/toggles`;
  } else {
    const res = await getProjectList<IProject[]>();
    const { success, data } = res;
    if (success && data) {
      if (data[0]) {
        const { key, environments } = data[0];
        let envKey = '';
        environments.some((environment: IEnvironment) => {
          if (environment.key) {
            envKey = environment.key;
            return true;
          }
          return environment;
        });
  
        if (key && envKey) {
          redirectUrl = `/${key}/${envKey}/toggles`;
        }
      } else {
        redirectUrl = '/projects';
      }
    }
  }

  return redirectUrl;
};