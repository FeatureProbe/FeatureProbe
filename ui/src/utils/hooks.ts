
import { useState } from 'react';

export const useLocalStorage = (key: string, initialValue: boolean | string) => {
  const [storedValue, setStoredValue] = useState(() => {
    if (typeof window === 'undefined') {
      return initialValue;
    }

    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.log(error);
      return initialValue;
    }
  });

  const setValue = (value: object) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);

      if (typeof window !== 'undefined') {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
      
    } catch (error) {
      console.log(error);
    }
  };
  return [storedValue, setValue];
};
