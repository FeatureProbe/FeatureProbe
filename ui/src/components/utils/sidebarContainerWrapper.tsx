import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
