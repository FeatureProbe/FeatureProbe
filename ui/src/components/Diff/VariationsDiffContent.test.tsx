import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import VariationsDiffContent from './VariationsDiffContent';
import { IntlWrapper } from 'components/utils/wrapper';
import idiff from './diff';

const content_a = idiff(
  [
    {
      name: 'name 1',
      value: 'value 1',
      description: 'description 1'
    },
    {
      name: 'name 2',
      value: 'value 2',
      description: 'description 2'
    },
    {
      name: 'name 3',
      value: 'value 3',
      description: 'description 3'
    },
  ],
  [
    {
      name: 'name 2',
      value: 'value 2',
      description: 'description 2'
    },
    {
      name: 'name 3',
      value: 'value 3',
      description: 'description 3 modify'
    },
    {
      name: 'name 4',
      value: 'value 4',
      description: 'description 4'
    },
  ]
);

const content_b = idiff(
  [
    {
      name: 'name 1',
      value: 'value 1',
      description: 'description 1'
    }
  ],
  [
    {
      nameErr: 'name 2',
      valueErr: 'value 2',
      descriptionErr: 'description 2'
    },
  ]
);

it('DiffVariations snapshot', (done) => {
  (async () => {
    const { asFragment, rerender } = render(<VariationsDiffContent content={content_a} />, {
      wrapper: IntlWrapper,
    });
    expect(asFragment()).toMatchSnapshot();

    rerender(<VariationsDiffContent content={content_b} />);
    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});
