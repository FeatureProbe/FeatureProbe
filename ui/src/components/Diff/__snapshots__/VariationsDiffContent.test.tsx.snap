// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffVariations snapshot 1`] = `
<DocumentFragment>
  <div
    class="diff-content"
  >
    <div
      class="before"
    >
      <div
        class="diff-fields"
      >
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class="table-header-name"
              >
                name
              </th>
              <th
                class=""
              >
                value
              </th>
              <th
                class=""
              >
                description
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item diff-item-remove"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  name 1
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  value 1
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  description 1
                </div>
              </td>
            </tr>
            <tr
              class="diff-item diff-item-same"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  name 2
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  value 2
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  description 2
                </div>
              </td>
            </tr>
            <tr
              class="diff-item"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  name 3
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  value 3
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  description 3
                </div>
              </td>
            </tr>
            <tr
              class="diff-item "
            >
              <td
                class="table-item-name"
              />
              <td
                class=""
              />
              <td
                class=""
              />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="after"
    >
      <div
        class="diff-fields"
      >
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class="table-header-name"
              >
                name
              </th>
              <th
                class=""
              >
                value
              </th>
              <th
                class=""
              >
                description
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item "
            >
              <td
                class="table-item-name"
              />
              <td
                class=""
              />
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item diff-item-same"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  name 2
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  value 2
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  description 2
                </div>
              </td>
            </tr>
            <tr
              class="diff-item"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  name 3
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  value 3
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  description 3 modify
                </div>
              </td>
            </tr>
            <tr
              class="diff-item diff-item-add"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  name 4
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  value 4
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-undefined"
                >
                  description 4
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffVariations snapshot 2`] = `
<DocumentFragment>
  <div
    class="diff-content"
  >
    <div
      class="before"
    >
      <div
        class="diff-fields"
      >
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class="table-header-name"
              >
                name
              </th>
              <th
                class=""
              >
                value
              </th>
              <th
                class=""
              >
                description
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item"
            >
              <td
                class="table-item-name"
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  name 1
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  value 1
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  description 1
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="after"
    >
      <div
        class="diff-fields"
      >
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class="table-header-name"
              >
                name
              </th>
              <th
                class=""
              >
                value
              </th>
              <th
                class=""
              >
                description
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item"
            >
              <td
                class="table-item-name"
              >
                null
              </td>
              <td
                class=""
              >
                null
              </td>
              <td
                class=""
              >
                null
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
